fe/fe-core/src/main/java/com/starrocks/connector/hive/HiveConnectorScanRangeSource.java

```diff
@@ -0,0 +1,409 @@
+                    String host = blockDesc.get().getDataNodeIp(hostId);
+                    TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress(host, -1));
+                    scanRangeLocations.addToLocations(scanRangeLocation);
+                }
+            } else {
+                TScanRangeLocation scanRangeLocation = new TScanRangeLocation(new TNetworkAddress("-1", -1));
+                scanRangeLocations.addToLocations(scanRangeLocation);
+            }
+            buffer.add(scanRangeLocations);
+        }
+
+    }
+
+    public Iterator<TScanRangeLocations> createScanRangeIterator(RemoteFileInfo remoteFileInfo) {
+        return new ScanRangeIterator(remoteFileInfo);
+    }
+
+    private void updateBackendSplitFile(RemoteFileInfo partition) {
+        ConnectContext connectContext = ConnectContext.get();
+        if (connectContext == null) {
+            return;
+        }
+        long hugeFileSize = connectContext.getSessionVariable().getConnectorHugeFileSize();
+        for (RemoteFileDesc fileDesc : partition.getFiles()) {
+            if (fileDesc.isSplittable()) {
+                backendSplitCount += (fileDesc.getLength() + hugeFileSize - 1) / hugeFileSize;
+            } else {
+                backendSplitCount += 1;
+            }
+        }
+        // if splits is small comparing to nodes, then better not let backend do split.
+        int nodes = connectContext.getAliveComputeNumber() + connectContext.getAliveBackendNumber();
+        backendSplitFile = (backendSplitCount > 2 * nodes);
+    }
+
+    private void updateIterator() {
+        if (!hasMoreOutput) {
+            return;
+        }
+        if (remoteFileInfoSource == null) {
+            init();
+        }
+        while ((iterator == null || !iterator.hasNext())) {
+            while (true) {
+                if (!remoteFileInfoSource.hasMoreOutput()) {
+                    hasMoreOutput = false;
+                    return;
+                } else {
+                    buffer = remoteFileInfoSource.getOutput();
+                }
+                if (buffer != null && buffer.getFiles() != null && buffer.getFiles().size() > 0) {
+                    break;
+                }
+            }
+            updateBackendSplitFile(buffer);
+            iterator = createScanRangeIterator(buffer);
+        }
+    }
+
+    @Override
+    public List<TScanRangeLocations> getOutputs(int maxSize) {
+        List<TScanRangeLocations> res = new ArrayList<>();
+        updateIterator();
+        while (hasMoreOutput) {
+            while (res.size() < maxSize && iterator.hasNext()) {
+                res.add(iterator.next());
+            }
+            if (res.size() >= maxSize) {
+                break;
+            }
+            updateIterator();
+        }
+        Collections.shuffle(res);
+        return res;
+    }
+
+    @Override
+    public boolean hasMoreOutput() {
```

Comment:
This is used for incremental scheduling, right?

Reply:
yes